### PR TITLE
[Merged by Bors] - Support raw buffers in AsBindGroup macro

### PR DESCRIPTION
**Pull request by AndrewB330**
# Objective

There was PR that introduced support for storage buffer is `AsBindGroup` macro [#6129](https://github.com/bevyengine/bevy/pull/6129), but it does not give more granular control over storage buffer, it will always copy all the data no matter which part of it was updated. There is also currently another open PR #6669 that tries to achieve exactly that, it is just not up to date and seems abandoned (Sorry if that is not right). In this PR I'm proposing a solution for both of these approaches to co-exist using `#[storage(n, buffer)]` and `#[storage(n)]` to distinguish between the cases.

We could also discuss in this PR if there is a need to extend this support to DynamicBuffers as well.

### CODE REVIEW

**Comment by IceSentry** [comment]
Instead of adding a parameter to `[storage(i)]`, I think you should introduce a new `[buffer(i)]` for the `Buffer` type. For the dynamic case it would just be `[buffer(i, dynamic)]`

**Comment by AndrewB330** [comment]
Actually I think keeping it inside `storage` attribute makes more sense. We have `texture` attribute for textures, `sampler` for samplers, `uniform` for uniform bindings and now `storage` for storage bindings. And in our case no matter what it is `Vec<Color>` or `Buffer` it will be considered as a storage binding from within the shader. So I think we should keep everything that binds to `var<storage>` inside `storage` attribute, maybe the scheme that I proposed is not optimal, I agree. We could come up with more verbose scheme, like `buffer_type=Raw\Dynamic\Auto` or something along those lines. 

Or maybe we could even detect which type was used under the hood of macro and decide what code to generate depending on the field type automatically? Not sure how do that properly, as I'm not very familiar with rust_ast and how to properly compare types.

**Comment by IceSentry** [comment]
Yeah, you're absolutely right actually. I'm not sure what I was thinking about last night.

File: crates/bevy_render/macros/src/as_bind_group.rs [hunk 1]
```diff
@@ -200,25 +200,40 @@ pub fn derive_as_bind_group(ast: syn::DeriveInput) -> Result<TokenStream> {
...
-                        #render_path::render_resource::OwnedBindingResource::Buffer(render_device.create_buffer_with_data(
-                            &#render_path::render_resource::BufferInitDescriptor {
-                                label: None,
-                                usage: #render_path::render_resource::BufferUsages::COPY_DST | #render_path::render_resource::BufferUsages::STORAGE,
-                                contents: buffer.as_ref(),
-                            },
-                        ))
-                    }});
+                    let min_binding_size = if buffer {
+                        quote! {None}
```
Comment (IceSentry) [inline crates/bevy_render/macros/src/as_bind_group.rs:212]: Does `Buffer` not implement `ShaderType`?

**Review by IceSentry — APPROVED** [review]

File: crates/bevy_render/macros/src/as_bind_group.rs (hunk 1 above)
Comment (AndrewB330) [reply crates/bevy_render/macros/src/as_bind_group.rs:212]: No, it doesn't implement it

**Comment by AndrewB330** [comment]
@alice-i-cecile what are the next steps for this PR? Should someone else review it? Thanks

**Comment by alice-i-cecile** [comment]
Yep, next step is to find a second reviewer with rendering expertise (so, not me). If they approve it, it'll get the `Ready-For-Final-Review` label, and then one of the maintainers will do a final review and merge it.

**Comment by AndrewB330** [comment]
@mockersf @superdump could I ask you for a review? Thanks!

**Review by mockersf — APPROVED** [review]

**Pull request closed without merging**